--- Day 2: Cube Conundrum ---

You're launched high into the atmosphere! The apex of your trajectory just barely reaches the surface of a large island floating in the sky. You gently land in a fluffy pile of leaves. It's quite cold, but you don't see much snow. An Elf runs over to greet you.

The Elf explains that you've arrived at Snow Island and apologizes for the lack of snow. He'll be happy to explain the situation, but it's a bit of a walk, so you have some time. They don't get many visitors up here; would you like to play a game in the meantime?

As you walk, the Elf shows you a small bag and some cubes which are either red, green, or blue. Each time you play this game, he will hide a secret number of cubes of each color in the bag, and your goal is to figure out information about the number of cubes.

To get information, once a bag has been loaded with cubes, the Elf will reach into the bag, grab a handful of random cubes, show them to you, and then put them back in the bag. He'll do this a few times per game.

You play several games and record the information from each game (your puzzle input). Each game is listed with its ID number (like the 11 in Game 11: ...) followed by a semicolon-separated list of subsets of cubes that were revealed from the bag (like 3 red, 5 green, 4 blue).

For example, the record of a few games might look like this:

Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green

In game 1, three sets of cubes are revealed from the bag (and then put back again). The first set is 3 blue cubes and 4 red cubes; the second set is 1 red cube, 2 green cubes, and 6 blue cubes; the third set is only 2 green cubes.

The Elf would first like to know which games would have been possible if the bag contained only 12 red cubes, 13 green cubes, and 14 blue cubes?

In the example above, games 1, 2, and 5 would have been possible if the bag had been loaded with that configuration. However, game 3 would have been impossible because at one point the Elf showed you 20 red cubes at once; similarly, game 4 would also have been impossible because the Elf showed you 15 blue cubes at once. If you add up the IDs of the games that would have been possible, you get 8.

Determine which games would have been possible if the bag had been loaded with only 12 red cubes, 13 green cubes, and 14 blue cubes. What is the sum of the IDs of those games?

In [107]:
class Bag:
    def __init__(self, red=0, green=0, blue=0):
        self.red = red
        self.green = green
        self.blue = blue

    def __str__(self):
        return "Red: " + self.red.__str__() + " Green: " + self.green.__str__() + " Blue: " + self.blue.__str__()

class Round:
    def __init__(self, red=0, green=0, blue=0):
        self.red = red
        self.green = green
        self.blue = blue

    def has_more_than(self, bag):
        output = (self.red > bag.red) or (self.green > bag.green) or (self.blue > bag.blue)
        ##print("----Round:---- " + self.__str__() + " ----Bag:---- " + bag.__str__() + " ----Output:---- " + output.__str__())
        return output
    
    def __str__(self):
        return "Red: " + self.red.__str__() + " Green: " + self.green.__str__() + " Blue: " + self.blue.__str__()

class Game:
    def __init__(self, game_string):
        game_id_string, rounds_string = game_string.split(':')
        self.id = self.get_game_id(game_id_string)
        self.rounds = self.create_rounds(rounds_string)

    def get_game_id(self, game_id_string):
        return int(game_id_string.split(' ')[1])

    def create_rounds(self, rounds_string):
        rounds = []
        for round_string in rounds_string.split(';'):
            red, green, blue = 0, 0, 0
            for splited in round_string.split(','):
                value, color = splited.strip().split(' ')
                if color == 'red':
                    red = int(value)
                elif color == 'green':
                    green = int(value)
                elif color == 'blue':
                    blue = int(value)   
            round = Round(red, green, blue)
            rounds.append(round)
        return rounds
    
    def any_round_has_more_than(self, bag):
        for round in self.rounds:
            if round.has_more_than(bag):
                return True
        return False
    
    def __rounds__str__(self):
        output = ""
        for round in self.rounds:
            output += round.__str__() + "; "
        return output
    
    def __str__(self):
        return("Game ID: " + self.id.__str__() + " Rounds: " + self.__rounds__str__())
    
class GameManager:
    def __init__(self, game_strings):
        self.games = self.add_games(game_strings)
        self.validGames = []

    def add_games(self, game_strings):
        games = []
        for game_string in game_strings:
            games.append(Game(game_string))
        return games

    def check_games(self, bag):
        for game in self.games:
            if game.any_round_has_more_than(bag) is False:
                self.validGames.append(game)
    
    def sum_valid_game_ids(self):
        return sum(game.id for game in self.validGames)
    
    def print_valid_games(self):
        print("Valid Games:")
        for game in self.validGames:
            print(game)
    
    def print_result(self):
        print("Sum valid Games:" + self.sum_valid_game_ids().__str__())

Determine which games would have been possible if the bag had been loaded with only 12 red cubes, 13 green cubes, and 14 blue cubes. What is the sum of the IDs of those games?

In [108]:
with open('input.txt', 'r') as file:
    lines = file.readlines()

gameManager = GameManager(lines)
bag = Bag(12, 13, 14)
gameManager.check_games(bag)
gameManager.print_result()
gameManager.print_valid_games()

Sum valid Games:1867
Valid Games:
Game ID: 1 Rounds: Red: 5 Green: 1 Blue: 0; Red: 6 Green: 0 Blue: 3; Red: 9 Green: 0 Blue: 0; Red: 4 Green: 1 Blue: 1; Red: 0 Green: 1 Blue: 2; Red: 1 Green: 0 Blue: 2; 
Game ID: 2 Rounds: Red: 12 Green: 2 Blue: 9; Red: 8 Green: 0 Blue: 12; Red: 9 Green: 2 Blue: 1; Red: 8 Green: 2 Blue: 12; Red: 4 Green: 0 Blue: 5; Red: 10 Green: 1 Blue: 9; 
Game ID: 4 Rounds: Red: 3 Green: 0 Blue: 0; Red: 3 Green: 3 Blue: 1; Red: 5 Green: 8 Blue: 3; Red: 0 Green: 8 Blue: 2; 
Game ID: 5 Rounds: Red: 1 Green: 1 Blue: 8; Red: 1 Green: 7 Blue: 1; Red: 0 Green: 7 Blue: 5; 
Game ID: 6 Rounds: Red: 2 Green: 0 Blue: 8; Red: 0 Green: 4 Blue: 5; Red: 4 Green: 3 Blue: 5; Red: 3 Green: 0 Blue: 9; Red: 7 Green: 5 Blue: 9; Red: 6 Green: 5 Blue: 9; 
Game ID: 7 Rounds: Red: 0 Green: 3 Blue: 4; Red: 3 Green: 7 Blue: 3; Red: 7 Green: 4 Blue: 6; Red: 5 Green: 1 Blue: 4; Red: 2 Green: 7 Blue: 6; Red: 4 Green: 1 Blue: 4; 
Game ID: 10 Rounds: Red: 2 Green: 0 Blue: 5; Red: 3 Green: 5 Blue: 

Your puzzle answer was 1867.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---

The Elf says they've stopped producing snow because they aren't getting any water! He isn't sure why the water stopped; however, he can show you how to get to the water source to check it out for yourself. It's just up ahead!

As you continue your walk, the Elf poses a second question: in each game you played, what is the fewest number of cubes of each color that could have been in the bag to make the game possible?

Again consider the example games from earlier:

Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green

    In game 1, the game could have been played with as few as 4 red, 2 green, and 6 blue cubes. If any color had even one fewer cube, the game would have been impossible.
    Game 2 could have been played with a minimum of 1 red, 3 green, and 4 blue cubes.
    Game 3 must have been played with at least 20 red, 13 green, and 6 blue cubes.
    Game 4 required at least 14 red, 3 green, and 15 blue cubes.
    Game 5 needed no fewer than 6 red, 3 green, and 2 blue cubes in the bag.

The power of a set of cubes is equal to the numbers of red, green, and blue cubes multiplied together. The power of the minimum set of cubes in game 1 is 48. In games 2-5 it was 12, 1560, 630, and 36, respectively. Adding up these five powers produces the sum 2286.

For each game, find the minimum set of cubes that must have been present. What is the sum of the power of these sets?


In [116]:
class MinimalBag(Bag):  
    def __init__(self, red=0, green=0, blue=0):
        super().__init__(red, green, blue)

    def set_new_max_red(self, red):
        if red > self.red:
            self.red = red

    def set_new_max_green(self, green):
        if green > self.green:
            self.green = green
        
    def set_new_max_blue(self, blue):
        if blue > self.blue:
            self.blue = blue

class PowerGame(Game):
    def __init__(self, game_string):
        super().__init__(game_string)
        self.minimalBag = MinimalBag()
        self.gamePower = 0
    
    def set_minimal_bag(self, round):
        for round in self.rounds:
            self.minimalBag.set_new_max_red(round.red)
            self.minimalBag.set_new_max_green(round.green)
            self.minimalBag.set_new_max_blue(round.blue)

    def calculate_power(self):
        self.gamePower = self.minimalBag.red * self.minimalBag.green * self.minimalBag.blue
    
class PowerGameManager(GameManager):
    def __init__(self, game_strings):
        super().__init__(game_strings)

    def add_games(self, game_strings):
        games = []
        for game_string in game_strings:
            games.append(PowerGame(game_string))
        return games
        
    def check_game_powers(self):
        for game in self.games:
            game.set_minimal_bag(game.rounds)
            game.calculate_power()
    
    def print_game_powers(self):
        print("Game Powers:")
        for game in self.games:
            print("Game ID: " + game.id.__str__() + " Power: " + game.gamePower.__str__())
            
    def sum_game_powers(self):
        return sum(game.gamePower for game in self.games)
    
    def print_sum_game_powers(self):
        print("Sum Game Powers:" + self.sum_game_powers().__str__())
    


In [117]:
with open('input.txt', 'r') as file:
    lines = file.readlines()

gameManager = PowerGameManager(lines)
gameManager.check_game_powers()
gameManager.print_game_powers()
gameManager.print_sum_game_powers()

Game Powers:
Game ID: 1 Power: 27
Game ID: 2 Power: 288
Game ID: 3 Power: 1040
Game ID: 4 Power: 120
Game ID: 5 Power: 56
Game ID: 6 Power: 315
Game ID: 7 Power: 294
Game ID: 8 Power: 288
Game ID: 9 Power: 336
Game ID: 10 Power: 390
Game ID: 11 Power: 468
Game ID: 12 Power: 64
Game ID: 13 Power: 98
Game ID: 14 Power: 162
Game ID: 15 Power: 32
Game ID: 16 Power: 605
Game ID: 17 Power: 720
Game ID: 18 Power: 1904
Game ID: 19 Power: 1404
Game ID: 20 Power: 42
Game ID: 21 Power: 990
Game ID: 22 Power: 52
Game ID: 23 Power: 60
Game ID: 24 Power: 800
Game ID: 25 Power: 864
Game ID: 26 Power: 180
Game ID: 27 Power: 1512
Game ID: 28 Power: 864
Game ID: 29 Power: 324
Game ID: 30 Power: 48
Game ID: 31 Power: 900
Game ID: 32 Power: 78
Game ID: 33 Power: 56
Game ID: 34 Power: 27
Game ID: 35 Power: 630
Game ID: 36 Power: 224
Game ID: 37 Power: 480
Game ID: 38 Power: 1672
Game ID: 39 Power: 494
Game ID: 40 Power: 1980
Game ID: 41 Power: 20
Game ID: 42 Power: 2700
Game ID: 43 Power: 912
Game ID: 44 P

Your puzzle answer was 84538.

Both parts of this puzzle are complete! They provide two gold stars: **

At this point, you should return to your Advent calendar and try another puzzle.

If you still want to see it, you can get your puzzle input.

You can also [Shareon Twitter Mastodon] this puzzle.